### PR TITLE
feat: add binary exponentiation algorithm

### PR DESCRIPTION
**Pull request by PraneethJain**
#### Description of Change
Implementation of the binary exponentiation algorithm, which calculates the nth power of an integer in O(log(n)) time.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: math/binary_exponentiation.c [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+#include <stdio.h>   // for output
+#include <stdlib.h>  // for exit()
+
+/**
+ * @brief Determines a raised to the nth power via binary exponentiation.
+ * @param a - base
+ * @param n - exponent
+ * @return a raised to the nth power
+ * @warning
+ * `int` can overflow very quickly.
```
Comment (realstealthninja) [inline math/binary_exponentiation.c:21]: use a larger and unsigned type like uint64

Comment (PraneethJain) [reply math/binary_exponentiation.c]: The base can be negative so switched to int64_t for that, and switched to uint64_t for the exponent.

**Comment by PraneethJain** [comment]
Hey @realstealthninja, I've made the requested changes. Could you re review this?

File: math/binary_exponentiation.c [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+ * Calculating the nth power of a number requires O(n) time using the naive
+ * method. This algorithm optimizes it through exponentiation by squaring.
+ * https://cp-algorithms.com/algebra/binary-exp.html
+ * @author [Praneeth Jain](https://github.com/PraneethJain)
+ */
+
+#include <assert.h>  // for assert()
+#include <stdint.h>  // for int64 types
+#include <stdio.h>   // for output
+#include <stdlib.h>  // for exit()
```
Comment (realstealthninja) [inline math/binary_exponentiation.c]:
```suggestion
#include <assert.h>  /// for assert()
#include <stdint.h>  /// for int64 types
#include <stdio.h>   /// for output
#include <stdlib.h>  /// for exit()
```

File: math/binary_exponentiation.c [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+/**
+ * @brief Determines a raised to the nth power via binary exponentiation.
+ * @param a - base
+ * @param n - exponent
+ * @return a raised to the nth power
+ * @warning
+ * can overflow very quickly.
+ */
+int64_t binary_exponentiation(int64_t a, uint64_t n)
```
Comment (realstealthninja) [inline math/binary_exponentiation.c]:
```suggestion
int64_t binary_exponentiation(int64_t base, uint64_t exponent)
```

File: math/binary_exponentiation.c [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+{
+    int64_t res = 1;
+    while (n > 0)
```
Comment (realstealthninja) [inline math/binary_exponentiation.c]:
```suggestion
    while (exponent > 0)
```

File: math/binary_exponentiation.c [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        if (n % 2 == 1)  // If the current bit is 1
```
Comment (realstealthninja) [inline math/binary_exponentiation.c]:
```suggestion
        if (exponent % 2 == 1)  // If the current bit is 1
```

File: math/binary_exponentiation.c [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        {
+            res *= a;  // Then it must be multiplied to the result
```
Comment (realstealthninja) [inline math/binary_exponentiation.c]:
```suggestion
            result *= base;  // Then it must be multiplied to the result
```

File: math/binary_exponentiation.c [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        a *= a;   // Square the base
```
Comment (realstealthninja) [inline math/binary_exponentiation.c]:
```suggestion
        base *= base;   // Square the base
```

File: math/binary_exponentiation.c [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        n >>= 1;  // Move to the next bit
```
Comment (realstealthninja) [inline math/binary_exponentiation.c]:
```suggestion
        exponent >>= 1;  // Move to the next bit
```

File: math/binary_exponentiation.c [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    }
+
+    return res;
```
Comment (realstealthninja) [inline math/binary_exponentiation.c]:
```suggestion
    return result;
```

File: math/binary_exponentiation.c [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+ * @brief Determines a raised to the nth power via binary exponentiation.
+ * @param a - base
+ * @param n - exponent
+ * @return a raised to the nth power
+ * @warning
+ * can overflow very quickly.
+ */
+int64_t binary_exponentiation(int64_t a, uint64_t n)
+{
+    int64_t res = 1;
```
Comment (realstealthninja) [inline math/binary_exponentiation.c]:
```suggestion
    int64_t result = 1;
```

File: math/binary_exponentiation.c [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+ */
+
+#include <assert.h>  // for assert()
+#include <stdint.h>  // for int64 types
+#include <stdio.h>   // for output
+#include <stdlib.h>  // for exit()
+
+/**
+ * @brief Determines a raised to the nth power via binary exponentiation.
+ * @param a - base
```
Comment (realstealthninja) [inline math/binary_exponentiation.c]:
```suggestion
 * @param base of the given number
```

File: math/binary_exponentiation.c [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+ * @param n - exponent
```
Comment (realstealthninja) [inline math/binary_exponentiation.c]:
```suggestion
 * @param exponent of the given number
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]
Great work i recommend adding more descriptive variables; 
yes a and n are the common math notation but it increases readability imo. 
but its completely fine if you prefer a^n notation lmk

File: math/binary_exponentiation.c (hunk 2 above)
Comment (realstealthninja) [reply math/binary_exponentiation.c]: doxygen needs thriple slashes to recognize this as a doc string for the includes

**Review by realstealthninja — APPROVED** [review]
Thank you so much for contributing!!! :heart:  the code is very elegant to read!

**Comment by PraneethJain** [comment]
Hey, could someone review this?

**Review by realstealthninja — APPROVED** [review]
requires review of code owners i'm afraid

File: math/binary_exponentiation.c [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+ * @return a raised to the nth power
+ * @warning
+ * can overflow very quickly.
+ */
+int64_t binary_exponentiation(int64_t base, uint64_t exponent)
+{
+    int64_t result = 1;
+    while (exponent > 0)
+    {
+        if (exponent % 2 == 1)  // If the current bit is 1
```
Comment (alexpantyukhin) [inline math/binary_exponentiation.c]: It's interesting for me. Is the following expression would be faster? Just thinking...
```
if (exponent & 1)
```

Comment (PraneethJain) [reply math/binary_exponentiation.c]: Right, I'll do that.

**Review by alexpantyukhin — APPROVED** [review]

**Comment by PraneethJain** [comment]
@Panquesito7

**Pull request closed without merging**